sklearn/pipeline.py

```diff
@@ -995,6 +995,12 @@ class FeatureUnion(TransformerMixin, _BaseComposition):
 
         .. versionadded:: 0.24
 
+    feature_names_in_ : ndarray of shape (`n_features_in_`,)
+        Names of features seen during :term:`fit`. Defined only when
+        `X` has feature names that are all strings.
```

Comment:
This will end up to be inconsistent with when `n_features_in_` is defined. `n_features_in_` is only defined if the first transformer in `transformer_list` exposes it. So there are two paths forward:

1. Always define `n_features_in_` **and** `feature_names_in_`
2. Have `feature_names_in_` also depend on if `transformer_list` defines it.

Option 2 will require less discussion because that is the status quo. For example, `Pipeline` also exposes `feature_names_in_` based on the first `step`:

https://github.com/scikit-learn/scikit-learn/blob/fba66ded6c17023f3748d3cd2d49630fb7abbb28/sklearn/pipeline.py#L787-L791

So for this PR, I recommend going with option 2 since it is already an improvement.

Reply:
If I add the following to `FeatureUnion`

```python
@property
def feature_names_in_(self):
    """Names of features seen during :term:`fit`."""
    # X is passed to all transformers so we just delegate to the first one
    return self.transformer_list[0][1].feature_names_in_
```

It works as expected. But if I instantiate `FeatureUnion` with a fitted transformer,

```python
scaler = StandardScaler()
scaler.fit(X_train)
union = FeatureUnion([("scale", scaler)])
assert hasattr(union, "feature_names_in_")
assert_array_equal(X_train.columns, union.feature_names_in_)
```

This will pass. Is this what we'd expect?

Reply:
Yea it is expected to pass, which is consistent with `Pipeline`. Your code snippet would be a good test to include in this PR.

Historically, `Pipeline` and `FeatureUnion` were developed before we were strict about scikit-learn's API conventions on "fitted attributes". There has been work to fix this: https://github.com/scikit-learn/scikit-learn/pull/8350, but it has been blocked by some real uses cases the current behavior enables.

Reply:
Got it.

- [ ] remove `self._check_feature_names(X, reset=True)` from the `FeatureUnion`'s `.fit(...)` method
- [ ] add the `feature_names_in_` property to `FeatureUnion`
- [ ] update `test_feature_union_feature_names_in_()` to include a check when `FeatureUnion` is instantiated with a previously fitted transformer